### ISSUE TITLE
subprocess does not persist the effects of setup-env.sh

### ISSUE THREAD

**Issue by qwofford** [LLNL/spack#24260]
@scottwittenburg  @becker33 

This subprocess call is problematic if you are using a custom spack env:
https://github.com/spack/spack/blob/5926056f3a4c78a65d41688cdde825d89ed2ac2e/lib/spack/spack/cmd/ci.py#L484

By introducing the `-l` argument, the results of `setup-env.sh` are cleared at runtime. To reproduce, try this:

```
-bash-4.2$ . <my project path>/spack/share/spack/setup-env.sh
-bash-4.2$ which spack
<my project path>/spack/bin/spack
-bash-4.2$ cat test.py
import subprocess
subprocess.Popen(['bash','-l','./test.sh'])
-bash-4.2$ cat test.sh 
#!/bin/bash
which spack
-bash-4.2$ python test.py 
-bash-4.2$ /home/qwofford/spack/bin/spack
```
Then try:
```
-bash-4.2$ which spack
<my project path>/spack/bin/spack
-bash-4.2$ cat test.py
import subprocess
subprocess.Popen(['bash','./test.sh'])
-bash-4.2$ cat test.sh
#!/bin/bash
which spack
-bash-4.2$ python test.py
-bash-4.2$ <my project path>/spack/bin/spack
```

### Error Message

I got a validator error from the version of spack that was in my PATH, because the version in my PATH was older than the version in my activated environment via `setup-env.sh`

**Comment by qwofford** [comment]
To expand on this a bit, I have `SPACK_ROOT` set in my `~/.bashrc`, which over-writes the effects of `./setup-env.sh`. This is confusing in the CI job log because I'll see:

```
which spack
/the/right/path/spack
gitlab ci rebuild
...
ERROR from /home/qwofford/spack
```

I had to peek into the spack source to find the subprocess call with the `-l` option. An alternative could be to remove SPACK_ROOT from my login scripts, but I think I shouldn't have to. I'm open to discussion on that, of course.

**Comment by qwofford** [comment]
I am using a Jacamar Gitlab runner

**Comment by qwofford** [comment]
Curious if this issue and the associated PR #24261 might make it into develop, or if removing `-l` might break something unexpected.

**Comment by alalazo** [comment]
Fixed in #31771
